src/main/java/org/folio/oaipmh/helpers/AbstractGetRecordsHelper.java

```diff
@@ -67,17 +66,17 @@
     return completedFuture(buildResponseWithErrors(oaipmh));
   }
 
-  private CompletableFuture<MetadataType> getOaiMetadata(HttpClientInterface httpClient, Request request, String id) {
+  private CompletableFuture<MetadataType> getOaiMetadata(Context ctx, HttpClientInterface httpClient, Request request, String id) {
     try {
       String metadataEndpoint = storageHelper.getMetadataEndpoint(id);
-      return httpClient.request(metadataEndpoint, request.getOkapiHeaders(), false)
-                       .thenApplyAsync(response -> buildOaiMetadata(request, response));
+      return VertxCompletableFuture.from(ctx, httpClient.request(metadataEndpoint, request.getOkapiHeaders(), false))
+                                   .thenApplyAsync(response -> buildOaiMetadata(request, response));
```

Comment:
When wrapping future with response to `VertxCompletableFuture` and calling `thenApplyAsync`, the logic will be run on the context i.e. on event-loop thread. 
So alternative solution is following:
```java
return httpClient.request(metadataEndpoint, request.getOkapiHeaders(), false)
                 .thenCompose(response -> VertxCompletableFuture.supplyBlockingAsync(ctx, () -> buildOaiMetadata(request, response)));
```

Reply:
In the first case, we perform construction of metadata records on the event loop (which in total might be long enough process). In other words we spend event-loop cycles to process single request.
With `supplyBlockingAsync` we run metadata construction logic on the worker thread(s) and free up even-loop for processing other requests that might be more efficient/correct way.